Update to v8.00

aaronbell opened the pull request:
This PR represents a pretty major overhaul of the font repro. 

Montserrat's development appears to have been stalled and forked, so this PR is a hope to help re-merge the varying repros. 

Previously the googlefonts repro was converted to UFOs and VTT hand-hinting added to the font (thanks Mike!). However, when rebuilding the font using the latest version of fontmake, I discovered that the hints were not re-integrated into the build process, and when they were, were not compatible with the UFO-generated font. As the amount of work to repair the VTT source would be the roughly same for the UFO-process as it would be for any other font source, I decided that instead of continuing to use the UFO sources, to convert the entire repro to the UFR format using Julieta's upstream source (https://github.com/JulietaUla/Montserrat) which had progressed beyond the UFO sources (including some bug fixes). 

The Montserrat project is now building from her most up-to-date source, using the UFR process, and hand-hints are being merged in as a post-processing step with VTT compilation via vttlib. The hints were updated for the new source and problematic glyphs corrected. All of the old UFOs and related build content have thus been removed from the repro. 

While regression testing the new version of the font versus the latest googlefonts version, I found *a lot* of changes. I tried my best to determine which are due to progression in the source from the last version, which are non-problematic modifications due to the change in build process, and which are new bugs that need to be fixed. I resolved all of the items I identified as problematic and left everything else, optimistic that since the previous version had been hinted that the glyph updates were signed off on. 😨 

In any case, please review the new version and if you approve, I'll push it upstream to Julieta's repro.

davelab6 committed:
Rename AUTHOR.txt to AUTHORS.txt

juandelperal committed:
Updated to 7.210

solmatas committed:
Changes on the Serbian locl. Bump version 7.22

We updated the Serbian glyphs in the upright and italic versions.

solmatas committed:
Added Latin Small Letter N with Circumflex Below

(including the uppercase, small case and the alternates)
ṋ
ncircumflexbelow.sc.ss01
ncircumflexbelow.sc
Ṋ
Ncircumflexbelow.ss01

solmatas committed:
Bump version 7.221

solmatas committed:
Added Navajo letterforms

juandelperal committed:
added Bitcoin, change version

juandelperal committed:
generated fonts, rm css for ttf and otf

aaronbell committed:
Updating Montserrat to the UFR format

- Does not build the "Alternates" version (will likely need to switch to an instance export to implement. Maybe even Glyphs? Or skip)
- Needs hinting to be updated and imported

aaronbell committed:
Now merges in hinting code

- still need to fix up the hinting code

Building alternates as well seems to be untenable from a single Glyphs source.  Likely will need to have a separate Glyphs file for that.

aaronbell committed:
Updated

- hinting now integrated. A few hints became broken, but were easily resolved. Now compiles without issue in vttlib

- has some fixes found during regression testing (like removing _center anchor from dotaccent)

-still needs regression testing to be finalized and approved.

aaronbell committed:
Regression tested and fonts rebuild w/ static instances

There are *a lot* of changed to this new version versus the previous version of Montserrat. It is hard to determine if they are desired. However, as it seems the work was previously approved (it would have to have been for the font to be hand hinted), I am optimistic that they're OK.

aaronbell commented:
@googlebot I consent.

aaronbell commented:
Rebased the font to fix the merge. 

One other note I forgot to mention previously — the current build process does not allow for the production of Montserrat Alternates and they have been removed. Hopefully in the future, if that variant is important, we can re-enable export.

aaronbell commented:
@mikedug Thanks for the review Mike! I'm actually really surprised to see that the offsets are not being positioned correctly. The components are properly placed in the source and export TTF, but it seems that the Light Latin Autohinter is interpreting them incorrectly. I wouldn't expect it, but maybe there's something wrong with my VTT install? 

If you run the autohinter on the output variable fonts, are these components being positioned correctly? If so, could you run it on both the Roman and Italic and send them over? I can merge it in.

mikedug commented:
I do not think there is anything wrong with your VTT install.

It is a known issue, the SVTCA[Y], output from the Light Latin Autohinter,  uses offsets that are interpreted incorrectly in Variable fonts. This a manual fix, which is what I did in my sources, for all composites that need it. In this example, the code should read

CALL[], 31, 13, 0, 0, 0, 86

31 is the point on the outline, and the third 0 (zero) is the offset from 31 on the component and 13, (or any point on the baseline) on the base glyph, which is zero in both cases. In the auto generated code, point 32 is being used for the offset, which is ok for the Light default, but not the black. You can have a look at my sources for how I have done this. 

It is a manual fix for a lot of glyphs.

aaronbell commented:
Fascinating. I've never run into this issue before on any variable font I worked on. 😮 

Is there a specific scenario that causes it to emerge? It would be nice to be able to implement this fix automatically (or better yet, avoid it entirely!).

mikedug commented:
if you open the glyph set in VTT, set the size to a higher size, 40ppem, or higher, use ctrl g to toggle between hinted and un-hinted, you will see many of the glyphs that have problems. Proofing any of these in the main Windows and Variation window, with the Variation Window set to show Black Variant, will quickly show the issue. It is also rarely just one instance that will have a problem

the function 86 uses a distance between a point on the accent and a point on the base glyph. When this distance is constant across the variation space, that is ok, but when that distance is different, which it often is, in variable font design, then there are very bad distortions, which simple proofing will show. 

This is clearly a bug, but I do not know of any quick or automatic fix. 

If you used VTT for other variable fonts, I strongly suggest you check that issue, by proofing in the glyph set for the Variation instances that are in the font, and using the grid-fit no grid-fit, (ctrl g) technique.

aaronbell committed:
Updated hinting and build process

Re-autohinting the fonts was causing problems as it resulted in losing all of Mike's manual adjustments to the TSI1 table. This was further compounded by the "flattenComponents" filter, which further changed OFFSET versus the hinted version.

Here, I've reverted back to Mike's original TSI1 code, modifying the new places where things weren't working correctly, and disabled flattenComponents to make sure component bits stay the same.

aaronbell committed:
Cleared out the old bits

Fonts rebuilt.

aaronbell commented:
@mikedug Thanks again for your help. I figured that I could preserve your exisiting `TSI` tables (with correcting the OFFSET value) if I disabled the "flattenComponents" filter. 

Unfortunately, this causes a FAIL in FontBakery, but it gets the font working for now. Someday when that filter is re-enabled, the `TSI1` table will need to be rebuilt. Hopefully at that point SVTCA[Y] will be fixed. 

Anyway, I reviewed the hinting and didn't find any further issues. Thanks!

aaronbell committed:
Update hinting.py

MatthewRFennell commented:
This is a lovely font, and this PR makes it much easier to build the fonts from source using free software on Linux, which would be a big help when packaging the font for software distributions.
Is there anything in particular holding back this PR from being merged at the moment? If so, is there anything I can do to help?

solmatas commented:
@googlebot I consent.

juandelperal commented:
Hey @aaronbell,

Thanks a lot for this contribution! 
We are resuming the work with Montserrat and taking a look at this branch and this seems to be the latest version.
Is there anything else needed to get it merged?

davelab6 commented:
@aaronbell @m4rc1e @mikedug does anyone know why this isn't merged?

juandelperal commented:
We started from here and added many improvements (v9) here:
https://github.com/JulietaUla/Montserrat/tree/merge-forked-changes
